sharedbuffers/mapped_struct.py

```diff
@@ -458,34 +682,102 @@ def __getitem__(self, index):
+                elif op == Py_NE:
+                    return True
+                else:
+                    raise NotImplementedError
+
+            _cmp = proxied_list_cmp(self, other)
+            if op == Py_EQ:
+                return _cmp == 0
+            elif op == Py_NE:
+                return _cmp != 0
+            elif op == Py_LT:
+                return _cmp < 0
+            elif op == Py_LE:
+                return _cmp <= 0
+            elif op == Py_GT:
+                return _cmp > 0
+            else: # op == Py_GE:
+                return _cmp >= 0
+    else:
+        def __cmp__(self, other):
+            if not islist(other):
+                raise NotImplementedError
+            return proxied_list_cmp(self, other)
+
+        def _ne(self, other):
+            if not islist(other):
+                return True
+            return proxied_list_cmp(self, other) != 0
+
+        def _eq(self, other):
+            if not islist(other):
                 return False
-        return True
+            return proxied_list_cmp(self, other) == 0
+
+    @cython.ccall
+    @cython.locals(offs = cython.ulonglong, dcode = cython.char, pbuf = 'const char *', objlen = cython.ulonglong)
+    def _cobjlen(self):
+
+        offs = self.offs
+        pbuf = cython.cast(cython.p_char, self.pybuf.buf)
+        dcode = pbuf[offs]
+
+        if dcode in ('B','b','H','h','I','i'):
+            objlen = cython.cast(cython.p_uint, pbuf + offs)[0] >> 8
+            if objlen == 0xFFFFFF:
+                objlen = cython.cast(cython.p_longlong, pbuf + offs + 4)[0]
+
+        elif dcode in ('q', 'd', 't'):
+            objlen = cython.cast(cython.p_longlong, pbuf + offs)[0] >> 8
 
-    def __eq__(self, other):
-        if cython.compiled:
-            if isinstance(self, proxied_list):
-                return proxied_list._eq(cython.cast(proxied_list, self), other)
-            elif isinstance(other, proxied_list):
-                return proxied_list._eq(cython.cast(proxied_list, other), self)
-            else:
-                raise TypeError
         else:
-            return proxied_list._eq(self, other)
+            raise ValueError("Inconsistent data, unknown type code %r" % (dcode,))
+
+        return objlen
+
+    def _objlen(self):
+        offs = self.offs
+
+        buf = self.buf
+        dcode = buf[offs]
+
+        if dcode in ('B','b','H','h','I','i'):
+            objlen, = struct.unpack('<I', buf[offs+1:offs+4] + '\x00')
+            if objlen == 0xFFFFFF:
+                objlen = struct.unpack_from('<Q', buf, offs + 4)
+
+        elif dcode in ('q', 'd', 't'):
+            objlen, = struct.unpack('<Q', buf[offs+1:offs+8] + '\x00')
+        else:
+            raise ValueError("Inconsistent data, unknown type code %r" % (dcode,))
+
+        return objlen
 
     def __len__(self):
-        return self.objlen
+        if cython.compiled:
+            return self._cobjlen()
+        else:
+            return self._objlen()
```

Comment:
Really... why do you copy methods around instead of making the implementation inside conditional?